components/suggest/src/yelp.rs

```diff
@@ -142,69 +153,219 @@ impl<'a> SuggestDao<'a> {
+    ) -> Result<Option<(String, usize)>> {
+        if query_words.is_empty() {
+            return Ok(None);
+        }
+
+        for n in (1..=std::cmp::min(MAX_MODIFIER_WORDS_NUMBER, query_words.len())).rev() {
+            let mut candidate_chunks: Box<dyn Iterator<Item = &[&str]>> = match find_from {
+                FindFrom::First => Box::new(query_words.chunks(n)),
+                _ => Box::new(query_words.rchunks(n)),
+            };
+            let candidate = candidate_chunks.next().unwrap_or(&[""]).join(" ");
+            if self.conn.query_row_and_then_cachable(
+                "
+                SELECT EXISTS (
+                    SELECT 1 FROM yelp_modifiers WHERE type = :type AND keyword = :word LIMIT 1
+                )
+                ",
+                named_params! {
+                    ":type": modifier_type,
+                    ":word": candidate.to_lowercase(),
+                },
+                |row| row.get::<_, bool>(0),
+                true,
+            )? {
+                return Ok(Some((candidate, n)));
+            }
+        }
+
+        Ok(None)
+    }
+
+    /// Find the subject for given query.
+    /// It returns Option<tuple> as follows:
+    /// (
+    ///   String: The keyword in DB (but the case is inherited by query).
+    ///   bool: Whether or not the keyword is exact match.
+    ///   usize: Number of words in query_words that match the keyword.
+    /// )
+    fn find_subject(&self, query_words: &[&str]) -> Result<Option<(String, bool, usize)>> {
+        if query_words.is_empty() {
+            return Ok(None);
+        }
+
+        let query_string = query_words.join(" ");
+
+        if let Ok(keyword_lowercase) = self.conn.query_row_and_then_cachable(
+            "SELECT keyword
+             FROM yelp_subjects
+             WHERE :query BETWEEN keyword AND keyword || x'FFFF'
+             ORDER BY LENGTH(keyword) ASC, keyword ASC
+             LIMIT 1",
+            named_params! {
+                ":query": query_string.to_lowercase(),
+            },
+            |row| row.get::<_, String>(0),
+            true,
+        ) {
+            let keyword = &query_string[0..keyword_lowercase.len()];
+            return Ok(Some((
+                keyword.to_string(),
+                true,
+                keyword.split_whitespace().count(),
+            )));
+        };
+
+        if query_string.len() < SUBJECT_PREFIX_MATCH_THRESHOLD {
+            return Ok(None);
+        }
+
+        if let Ok(keyword_lowercase) = self.conn.query_row_and_then_cachable(
+            "SELECT keyword
+             FROM yelp_subjects
+             WHERE keyword BETWEEN :query AND :query || x'FFFF'
+             ORDER BY LENGTH(keyword) ASC, keyword ASC
+             LIMIT 1",
+            named_params! {
+                ":query": query_string.to_lowercase(),
+            },
+            |row| row.get::<_, String>(0),
+            true,
+        ) {
+            let keyword = format!(
```

Comment:
Please add a comment here about why we're using `format!`, i.e. we want to preserve the query as the user typed it including its case.